In [2]:
import pandas as pd
import numpy as np

In [3]:
df_raw = pd.read_excel("../data/TABLA_ACCIDENTES_22.xlsx", sheet_name=0) #el excel tiene dos hojas pero solo queremos la primera

df_raw.head()

ID_ACCIDENTE  ANYO  MES  DIA_SEMANA  HORA  COD_PROVINCIA  COD_MUNICIPIO  \
0             1  2022    1           6     3              1              0   
1             2  2022    1           5    16              1           1059   
2             3  2022    1           6    21              1              0   
3             4  2022    1           7    12              1              0   
4             5  2022    1           7    12              1              0   

   ISLA  ZONA  ZONA_AGRUPADA  ... PRIORI_OTRA  CONDICION_NIVEL_CIRCULA  \
0   NaN     1              1  ...         999                        1   
1   NaN     1              1  ...         999                        1   
2   NaN     1              1  ...           0                        1   
3   NaN     1              1  ...         999                        1   
4   NaN     1              1  ...         999                        1   

   CONDICION_FIRME  CONDICION_ILUMINACION  CONDICION_METEO  CONDICION_NIEBLA  \
0                1                      6                1               NaN   
1                3                      1                3               NaN   
2                1                      6                2               NaN   
3                3                      1                3               NaN   
4                1                      1                1               NaN   

   CONDICION_VIENTO  VISIB_RESTRINGIDA_POR  ACERA  TRAZADO_PLANTA  
0               NaN                      1    998               3  
1               NaN                      1    998               2  
2               NaN                      1    998               2  
3               NaN                      1    998               1  
4               NaN                      1    998               3  

[5 rows x 73 columns]

In [4]:
#Esto es solo para no tener que cargar el dataframe cada vez que nos equivoquemos en algo, con ejecutar esta línea vale
df = df_raw

In [5]:
df.shape

(97916, 73)

In [6]:
df.duplicated().sum()

0

Hay hasta 31 columnas indicando la gravedad del accidente, por lo que lo primero que haremos será agruparlas todas en una donde tan solo se indique la gravedad como leve, grave o mortal.

Lo primero para ello es entender nuestros datos, por ejemplo, ¿la columna que indica los fallecidos en 30 días integra la columna fallecidos en 24 horas o son independientes?

In [7]:
df.loc[df['TOTAL_MU24H'] > df['TOTAL_MU30DF']]

Empty DataFrame
Columns: [ID_ACCIDENTE, ANYO, MES, DIA_SEMANA, HORA, COD_PROVINCIA, COD_MUNICIPIO, ISLA, ZONA, ZONA_AGRUPADA, CARRETERA, KM, SENTIDO_1F, TITULARIDAD_VIA, TIPO_VIA, TIPO_ACCIDENTE, TOTAL_MU24H, TOTAL_HG24H, TOTAL_HL24H, TOTAL_VICTIMAS_24H, TOTAL_MU30DF, TOTAL_HG30DF, TOTAL_HL30DF, TOTAL_VICTIMAS_30DF, TOTAL_VEHICULOS, TOT_PEAT_MU24H, TOT_BICI_MU24H, TOT_CICLO_MU24H, TOT_MOTO_MU24H, TOT_TUR_MU24H, TOT_FURG_MU24H, TOT_CAM_MENOS3500_MU24H, TOT_CAM_MAS3500_MU24H, TOT_BUS_MU24H, TOT_OTRO_MU24H, TOT_SINESPECIF_MU24H, TOT_PEAT_MU30DF, TOT_BICI_MU30DF, TOT_CICLO_MU30DF, TOT_MOTO_MU30DF, TOT_TUR_MU30DF, TOT_FURG_MU30DF, TOT_CAM_MENOS3500_MU30DF, TOT_CAM_MAS3500_MU30DF, TOT_BUS_MU30DF, TOT_VMP_MU30DF, TOT_OTRO_MU30DF, TOT_SINESPECIF_MU30DF, NUDO, NUDO_INFO, CARRETERA_CRUCE, PRIORI_NORMA, PRIORI_AGENTE, PRIORI_SEMAFORO, PRIORI_VERT_STOP, PRIORI_VERT_CEDA, PRIORI_HORIZ_STOP, PRIORI_HORIZ_CEDA, PRIORI_MARCAS, PRIORI_PEA_NO_ELEV, PRIORI_PEA_ELEV, PRIORI_MARCA_CICLOS, PRIORI_CIRCUNSTANCIAL, PRIORI_OTRA, CONDICION_NIVEL_CIRCULA, CONDICION_FIRME, CONDICION_ILUMINACION, CONDICION_METEO, CONDICION_NIEBLA, CONDICION_VIENTO, VISIB_RESTRINGIDA_POR, ACERA, TRAZADO_PLANTA]
Index: []

[0 rows x 73 columns]

In [8]:
#Hacemos lo mismo para el número total de víctimas
df.loc[df['TOTAL_VICTIMAS_24H'] > df['TOTAL_VICTIMAS_30DF']]

Empty DataFrame
Columns: [ID_ACCIDENTE, ANYO, MES, DIA_SEMANA, HORA, COD_PROVINCIA, COD_MUNICIPIO, ISLA, ZONA, ZONA_AGRUPADA, CARRETERA, KM, SENTIDO_1F, TITULARIDAD_VIA, TIPO_VIA, TIPO_ACCIDENTE, TOTAL_MU24H, TOTAL_HG24H, TOTAL_HL24H, TOTAL_VICTIMAS_24H, TOTAL_MU30DF, TOTAL_HG30DF, TOTAL_HL30DF, TOTAL_VICTIMAS_30DF, TOTAL_VEHICULOS, TOT_PEAT_MU24H, TOT_BICI_MU24H, TOT_CICLO_MU24H, TOT_MOTO_MU24H, TOT_TUR_MU24H, TOT_FURG_MU24H, TOT_CAM_MENOS3500_MU24H, TOT_CAM_MAS3500_MU24H, TOT_BUS_MU24H, TOT_OTRO_MU24H, TOT_SINESPECIF_MU24H, TOT_PEAT_MU30DF, TOT_BICI_MU30DF, TOT_CICLO_MU30DF, TOT_MOTO_MU30DF, TOT_TUR_MU30DF, TOT_FURG_MU30DF, TOT_CAM_MENOS3500_MU30DF, TOT_CAM_MAS3500_MU30DF, TOT_BUS_MU30DF, TOT_VMP_MU30DF, TOT_OTRO_MU30DF, TOT_SINESPECIF_MU30DF, NUDO, NUDO_INFO, CARRETERA_CRUCE, PRIORI_NORMA, PRIORI_AGENTE, PRIORI_SEMAFORO, PRIORI_VERT_STOP, PRIORI_VERT_CEDA, PRIORI_HORIZ_STOP, PRIORI_HORIZ_CEDA, PRIORI_MARCAS, PRIORI_PEA_NO_ELEV, PRIORI_PEA_ELEV, PRIORI_MARCA_CICLOS, PRIORI_CIRCUNSTANCIAL, PRIORI_OTRA, CONDICION_NIVEL_CIRCULA, CONDICION_FIRME, CONDICION_ILUMINACION, CONDICION_METEO, CONDICION_NIEBLA, CONDICION_VIENTO, VISIB_RESTRINGIDA_POR, ACERA, TRAZADO_PLANTA]
Index: []

[0 rows x 73 columns]

No existe ningun accidente con más muertes en 24 horas que en 30 días, se asume por tanto que las columnas en 30 días ya incluyen las de 24 horas. Vamos a ver si se aplica a los hospitalizados y no hospitalizados también.

No vamos a eliminar el resto de columnas sobre fallecidos ya que aunque no se vayan a incluir en nuestro modelo pueden ser interesantes para analizar estadísticas.

Otra cuestión parecida a la anterior es si el número de heridos hospitalizados incluye también a los que acabaron muriendo. Este problema es más complicado, ya que si una víctima es hospitalizada y acaba muriendo en los 30 días posteriores es posible que cuente dentro de los dos grupos, no obstante dado que el total de victimas se calcula con la suma de fallecidos, hospitalizados y no hospitalizados, en teoría si alguien acaba falleciendo en el hospital no debería contar como victima hospitalizada ya que sino al hacer la suma estarían contando doble las victimas en estos casos, pero vamos a comprobarlo.

In [9]:
df[['TOTAL_HG24H','TOTAL_MU30DF','TOTAL_HG30DF']].loc[df['TOTAL_MU30DF'] > df['TOTAL_HG30DF']]

TOTAL_HG24H  TOTAL_MU30DF  TOTAL_HG30DF
16               0             1             0
133              0             1             0
196              0             1             0
324              1             1             0
446              0             2             0
...            ...           ...           ...
97610            0             1             0
97619            0             1             0
97622            0             1             0
97646            1             1             0
97712            1             1             0

[1351 rows x 3 columns]

Como se puede comprobar hay columnas en las que el número de fallecidos en 30 días supera al de ingresados en el mismo periodo, aunque pudiera ser que fallecieron fuera sin ser ingresados esto sería creíble en los casos donde la víctima falleció en el propio accidente, pero filas como la 6 donde la víctima no falleció en 24 horas pero si en 30 días indican que en efecto la columna de fallecidos es independiente de la de hospitalizados en el mismo periodo.

In [10]:
#más hospitalizados en 24 horas que en 30 días
df[['TOTAL_MU24H','TOTAL_MU30DF','TOTAL_HG24H', 'TOTAL_HG30DF','TOTAL_VICTIMAS_24H',
       'TOTAL_VICTIMAS_30DF','TOTAL_HL24H','TOTAL_HL30DF']].loc[df['TOTAL_HG24H'] > df['TOTAL_HG30DF']]

TOTAL_MU24H  TOTAL_MU30DF  TOTAL_HG24H  TOTAL_HG30DF  \
324              0             1            1             0   
605              0             1            1             0   
620              0             1            4             3   
1568             0             1            1             0   
1763             0             1            1             0   
...            ...           ...          ...           ...   
97226            0             1            1             0   
97229            0             1            1             0   
97506            0             1            1             0   
97646            0             1            1             0   
97712            0             1            1             0   

       TOTAL_VICTIMAS_24H  TOTAL_VICTIMAS_30DF  TOTAL_HL24H  TOTAL_HL30DF  
324                     1                    1            0             0  
605                     1                    1            0             0  
620                     4                    4            0             0  
1568                    3                    3            2             2  
1763                    1                    1            0             0  
...                   ...                  ...          ...           ...  
97226                   1                    1            0             0  
97229                   1                    1            0             0  
97506                   1                    1            0             0  
97646                   1                    1            0             0  
97712                   1                    1            0             0  

[232 rows x 8 columns]

Es curioso que si que haya algunos casos donde hay más hospitalizados en 24 horas que en 30 días, por lo que a priori parece que en este caso estár hospitalizado en las primeras 24 horas no implica contar como que estuviste hospitalizado en los 30 días posteriores, no obstante hay algunas incoherencias:
- En todos los casos hubo un fallecido en 30 días y ninguno en 24 horas (Cuando solo era una victima)
- Si estuvo hospitalizado menos de 24 horas y no falleció en esas horas cabría pensar que se trata de un accidente leve, pero solo hay 232 casos cuando los accidentes leves deberían ser mayoritarios, además falleció en los siguientes 30 días lo que indica que fue un accidente mortal.

Teoría: Como hemos mencionado anteriormente alguien fallecido en 30 días no será incluido en la columna de hospitalizados en ese periodo ya que generaría un conteo doble en el total de víctimas, no obstante, si será incluido en los hospitalizados en 24 horas ya que eso no genera ningúna incoherencia. Por tanto, los casos que observamos en la tabla anterior son casos donde la victima del accidente fue llevada al hospital de inmediato (Hospitalizado 24 horas), no falleció en las primeras 24 horas pero sí falleció en los siguientes días y lo sacaron de la lista de hospitalizados en 30 días. En estos casos la víctima nunca fallece en las primeras 24 horas porque de hacerlo se le sacaría del número de hospitalizados en ese periodo (y en el siguiente también ya que contaría como fallecido).

Vamos a confirmar esta teoría, para ello se tienen que dar varias condiciones, que son:
- 1º Si el número de fallecidos en 24 horas es 0 y el número de fallecidos en 30 días es igual al de ingresados en 24 horas y mayor a 0, el número de ingresados en 30 días nunca podrá ser igual ni mayor al de fallecidos en 30 días.
- 2º En los casos donde el número de fallecidos en los dós periodos sea el mismo, y el número de ingresados también lo sea, serán siempre casos en donde hubo más de una víctima (al menos una que falleció y una que sobrevivió)
- 3º El total de víctimas en ambos periodos siempre será el mismo.

In [11]:
# 1º Condición
df[['TOTAL_MU24H','TOTAL_MU30DF','TOTAL_HG24H', 'TOTAL_HG30DF','TOTAL_VICTIMAS_24H',
       'TOTAL_VICTIMAS_30DF','TOTAL_HL24H','TOTAL_HL30DF']].loc[(df['TOTAL_MU24H'] == 0) & (df['TOTAL_MU30DF']==df['TOTAL_HG24H']) & (df['TOTAL_HG30DF']>=df['TOTAL_MU30DF']) & (df['TOTAL_MU30DF']>0)]

Empty DataFrame
Columns: [TOTAL_MU24H, TOTAL_MU30DF, TOTAL_HG24H, TOTAL_HG30DF, TOTAL_VICTIMAS_24H, TOTAL_VICTIMAS_30DF, TOTAL_HL24H, TOTAL_HL30DF]
Index: []

In [12]:
#2º Condición
df[['TOTAL_MU24H','TOTAL_MU30DF','TOTAL_HG24H', 'TOTAL_HG30DF','TOTAL_VICTIMAS_24H',
       'TOTAL_VICTIMAS_30DF']].loc[(df['TOTAL_MU24H'] == df['TOTAL_MU30DF']) & (df['TOTAL_HG24H']==df['TOTAL_MU30DF']) & (df['TOTAL_HG24H']==1)]

TOTAL_MU24H  TOTAL_MU30DF  TOTAL_HG24H  TOTAL_HG30DF  \
975              1             1            1             1   
1303             1             1            1             1   
1324             1             1            1             1   
1332             1             1            1             1   
1379             1             1            1             1   
...            ...           ...          ...           ...   
93982            1             1            1             1   
94088            1             1            1             1   
94451            1             1            1             1   
97596            1             1            1             1   
97787            1             1            1             1   

       TOTAL_VICTIMAS_24H  TOTAL_VICTIMAS_30DF  
975                     2                    2  
1303                    2                    2  
1324                    2                    2  
1332                    2                    2  
1379                    2                    2  
...                   ...                  ...  
93982                   2                    2  
94088                   4                    4  
94451                   2                    2  
97596                   2                    2  
97787                   3                    3  

[159 rows x 6 columns]

In [13]:
#3º Condición
df.loc[df['TOTAL_VICTIMAS_24H'] != df['TOTAL_VICTIMAS_30DF']]

Empty DataFrame
Columns: [ID_ACCIDENTE, ANYO, MES, DIA_SEMANA, HORA, COD_PROVINCIA, COD_MUNICIPIO, ISLA, ZONA, ZONA_AGRUPADA, CARRETERA, KM, SENTIDO_1F, TITULARIDAD_VIA, TIPO_VIA, TIPO_ACCIDENTE, TOTAL_MU24H, TOTAL_HG24H, TOTAL_HL24H, TOTAL_VICTIMAS_24H, TOTAL_MU30DF, TOTAL_HG30DF, TOTAL_HL30DF, TOTAL_VICTIMAS_30DF, TOTAL_VEHICULOS, TOT_PEAT_MU24H, TOT_BICI_MU24H, TOT_CICLO_MU24H, TOT_MOTO_MU24H, TOT_TUR_MU24H, TOT_FURG_MU24H, TOT_CAM_MENOS3500_MU24H, TOT_CAM_MAS3500_MU24H, TOT_BUS_MU24H, TOT_OTRO_MU24H, TOT_SINESPECIF_MU24H, TOT_PEAT_MU30DF, TOT_BICI_MU30DF, TOT_CICLO_MU30DF, TOT_MOTO_MU30DF, TOT_TUR_MU30DF, TOT_FURG_MU30DF, TOT_CAM_MENOS3500_MU30DF, TOT_CAM_MAS3500_MU30DF, TOT_BUS_MU30DF, TOT_VMP_MU30DF, TOT_OTRO_MU30DF, TOT_SINESPECIF_MU30DF, NUDO, NUDO_INFO, CARRETERA_CRUCE, PRIORI_NORMA, PRIORI_AGENTE, PRIORI_SEMAFORO, PRIORI_VERT_STOP, PRIORI_VERT_CEDA, PRIORI_HORIZ_STOP, PRIORI_HORIZ_CEDA, PRIORI_MARCAS, PRIORI_PEA_NO_ELEV, PRIORI_PEA_ELEV, PRIORI_MARCA_CICLOS, PRIORI_CIRCUNSTANCIAL, PRIORI_OTRA, CONDICION_NIVEL_CIRCULA, CONDICION_FIRME, CONDICION_ILUMINACION, CONDICION_METEO, CONDICION_NIEBLA, CONDICION_VIENTO, VISIB_RESTRINGIDA_POR, ACERA, TRAZADO_PLANTA]
Index: []

[0 rows x 73 columns]

Todas las condiciones se han cumplido, por lo que parece que nuestra teoría era correcta. Además se termina de concluir que las columnas de 30 días incluyen las de 24 horas.

In [14]:
# ALEX: Este creo que no aporta nada pero lo he dejado por si opinas lo contrario, si estás de acuerdo bórralo
df[['TOTAL_MU30DF','TOTAL_HG30DF','TOTAL_HL30DF','TOTAL_VICTIMAS_30DF']].loc[(df['TOTAL_MU30DF'] == df['TOTAL_HG30DF']) & (df['TOTAL_MU30DF'] > 1)]

TOTAL_MU30DF  TOTAL_HG30DF  TOTAL_HL30DF  TOTAL_VICTIMAS_30DF
12034             2             2             2                    6
12851             2             2             0                    4
29576             2             2             6                   10
30031             2             2             2                    6
31432             3             3             0                    6
83079             2             2             0                    4
88434             2             2             6                   10
91138             2             2             0                    4

In [15]:
#Aquí habría que ver si borramos la de no hospitalizados en 24 horas, nos sirve de algo?
df = df.drop(['TOTAL_MU24H','TOTAL_VICTIMAS_24H'], axis=1)

Diferenciaremos entre cuatro tipos de accidentes: Sin lesiones, leve, grave y mortal. Ya que en el dataset no se indica el tipo de lesión vamos a establecer el siguiente criterio:

- Mortal = La víctima fallece (en cualquier periodo).
- Leve = Fue al hospital
- Sin lesiones = No fue al hospital

In [16]:
# Crear la columna TYP_ACC basada en las condicione
def clasificar_accidentes(column):
    if column['TOTAL_MU30DF'] > 1:
        return 'Mortal'
    elif column['TOTAL_HG30DF'] > 1 & column['TOTAL_MU30DF'] < 0: #Menor a cero?
        return 'Leve'
    else: 
        return 'Sin_lesiones'

df['TYP_ACC'] = df.apply(clasificar_accidentes, axis=1)

#Creo que hay algo mal en el código, aparecen todos como Sin_lesiones, no lo he comprobado a fondo pero creo que es así,
# Probablemente sea lo de menor a 0 en lugar de igual a 0. Además los hospitalizados deberían ser mayor a 0 no mayor a 1.


Se establece la columna tipo de accidente en función de la prioridad:
    Si hay una victima mortal, el accidente se clasifica como mortal
    Si no hay ningun accindente mortal pero si ha sigo ingresado el accidente se clasifica como Leve
    Si no hay ningun accidnete mortal ni nadie ha sido ingresado el accidente se clasifica como Sin_lesiones

# Limpieza y simplificación del Dataset

In [17]:
# Eliminar irrelevantes
df = df.drop(['ID_ACCIDENTE','ZONA_AGRUPADA','ANYO','TITULARIDAD_VIA','TOT_PEAT_MU24H','TOT_BICI_MU24H','TOT_CICLO_MU24H', 
        'TOT_MOTO_MU24H', 'TOT_TUR_MU24H','TOT_FURG_MU24H','TOT_CAM_MENOS3500_MU24H','TOT_CAM_MAS3500_MU24H', 'TOT_BUS_MU24H',
       'TOT_OTRO_MU24H','TOT_SINESPECIF_MU24H','TOT_BUS_MU30DF','TOT_SINESPECIF_MU30DF',
       'NUDO_INFO','CARRETERA_CRUCE','CONDICION_NIEBLA','CONDICION_VIENTO','TOT_CAM_MENOS3500_MU30DF'], axis=1)


<font color="red">No entiendo a que se refiere la última columna del dataset 'TRAZADDO_PLANTA' la eliminamos? </font>

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97916 entries, 0 to 97915
Data columns (total 50 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   MES                      97916 non-null  int64  
 1   DIA_SEMANA               97916 non-null  int64  
 2   HORA                     97916 non-null  int64  
 3   COD_PROVINCIA            97916 non-null  int64  
 4   COD_MUNICIPIO            97916 non-null  int64  
 5   ISLA                     5207 non-null   float64
 6   ZONA                     97916 non-null  int64  
 7   CARRETERA                97916 non-null  object 
 8   KM                       50944 non-null  float64
 9   SENTIDO_1F               97916 non-null  int64  
 10  TIPO_VIA                 97916 non-null  int64  
 11  TIPO_ACCIDENTE           97916 non-null  int64  
 12  TOTAL_HG24H              97916 non-null  int64  
 13  TOTAL_HL24H              97916 non-null  int64  
 14  TOTAL_MU30DF          

Parece que el dataset no tiene valores nulos excepto por la columna 'ISLA' que es normal porque la mayoría de accidentes ocurrieron en el territorio peninsular y en 'KM' que se trata de carreteras no inventariadas. No obstante esto no es cierto ya que en el dataset se utiliza 999 para indicar los valores faltantes, por eso parece que no hay nulos, vamos a eliminar estos valores para ver realmente como es nuestro Dataframe. También vamos a sustituir los valores nulos en 'ISLA' con 0 para indicar que no aplica.

In [19]:
df['ISLA'] = df['ISLA'].fillna(0) #Rellenar columna ISLA

#Gestionar nulos en columnas prioridad
cols_prioridad = ['PRIORI_NORMA', 'PRIORI_AGENTE', 'PRIORI_SEMAFORO', 'PRIORI_VERT_STOP',
                  'PRIORI_VERT_CEDA', 'PRIORI_HORIZ_STOP', 'PRIORI_HORIZ_CEDA', 'PRIORI_MARCAS',
                  'PRIORI_PEA_NO_ELEV', 'PRIORI_PEA_ELEV', 'PRIORI_MARCA_CICLOS',
                  'PRIORI_CIRCUNSTANCIAL', 'PRIORI_OTRA']

df.loc[(df['PRIORI_NORMA'] == 999) & (df['NUDO'] == 1), cols_prioridad] = df.loc[(df['PRIORI_NORMA'] == 999) & (df['NUDO'] == 1), cols_prioridad].replace(999, np.nan)

df[['ACERA', 'VISIB_RESTRINGIDA_POR']] = df[['ACERA', 'VISIB_RESTRINGIDA_POR']].replace([999, 18], np.nan)


In [20]:
#Gestionar columna 'KM' muchos valores nulos, valores 0, 999, 9999 que también significan nulo

In [21]:
#Comprobar el número de valores nulos en las columnas ajustadas
df_subset_nulls = df[['ISLA'] + cols_prioridad + ['ACERA', 'VISIB_RESTRINGIDA_POR']]
df_subset_nulls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97916 entries, 0 to 97915
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ISLA                   97916 non-null  float64
 1   PRIORI_NORMA           91150 non-null  float64
 2   PRIORI_AGENTE          91150 non-null  float64
 3   PRIORI_SEMAFORO        91150 non-null  float64
 4   PRIORI_VERT_STOP       91150 non-null  float64
 5   PRIORI_VERT_CEDA       91150 non-null  float64
 6   PRIORI_HORIZ_STOP      91150 non-null  float64
 7   PRIORI_HORIZ_CEDA      91150 non-null  float64
 8   PRIORI_MARCAS          91150 non-null  float64
 9   PRIORI_PEA_NO_ELEV     91150 non-null  float64
 10  PRIORI_PEA_ELEV        91150 non-null  float64
 11  PRIORI_MARCA_CICLOS    91150 non-null  float64
 12  PRIORI_CIRCUNSTANCIAL  91150 non-null  float64
 13  PRIORI_OTRA            91150 non-null  float64
 14  ACERA                  94950 non-null  float64
 15  VI

Aunque hay más nulos que antes, siguen siendo pocos teniendo en cuenta el tamaño del dataset, por lo que no eliminaremos ninguna columna por este criterio.

In [22]:
#Combinar tipos de prioridad en una sola columna que determine el tipo de prioridad

Hay ciertas variables que aunque son categóricas como la provincia, el tipo de vía, el tipo de accidente, etc. Se encuentran en formato numérico, lo cual para entrenar el modelo será mejor pero para visualizar los datos puede ser muy confuso, por lo que vamos a crear diccionarios de todas ellas para ayudar visualizar los datos.

In [48]:
#Diccionarios

dia_semana = {1:'Lunes', 2:'Martes', 3:'Miercoles', 4:'Jueves', 5:'Viernes', 6:'Sábado', 7:'Domingo'}

cod_provincia = {1: 'Álava', 2: 'Albacete', 3: 'Alicante', 4: 'Almería', 5: 'Ávila', 6: 'Badajoz', 7: 'Islas Baleares', 8: 'Barcelona', 
                9: 'Burgos', 10: 'Cáceres', 11: 'Cádiz', 12: 'Castellón', 13: 'Ciudad Real', 14: 'Córdoba', 15: 'Coruña', 16: 'Cuenca', 
                17: 'Girona', 18: 'Granada', 19: 'Guadalajara', 20: 'Gipuzkoa', 21: 'Huelva', 22: 'Huesca', 23: 'Jaén', 24: 'León', 
                25: 'Lleida', 26: 'Rioja, La', 27: 'Lugo', 28: 'Madrid', 29: 'Málaga', 30: 'Murcia', 31: 'Navarra', 32: 'Ourense', 
                33: 'Asturias', 34: 'Palencia', 35: 'Las Palmas', 36: 'Pontevedra', 37: 'Salamanca', 38: 'Santa Cruz de Tenerife', 
                39: 'Cantabria', 40: 'Segovia', 41: 'Sevilla', 42: 'Soria', 43: 'Tarragona', 44: 'Teruel', 45: 'Toledo', 46: 'Valencia', 
                47: 'Valladolid', 48: 'Bizkaia', 49: 'Zamora', 50: 'Zaragoza', 51: 'Ceuta', 52: 'Melilla'}

#Los códigos de municipios se encuentran en un excel por lo que primero hay que abrirlos y después crear el diccionario
df_municipios = pd.read_excel('../data/diccionario_municipio.xlsx', header=0, converters= {
    'COD': lambda x: str(x) #Convertir a texto para que no se eliminen los 0
})
cod_municipio = dict(zip(df_municipios['COD'], df_municipios['NOMBRE '])) #crear diccionario

In [ ]:
# https://www.ine.es/dyngs/INEbase/es/operacion.htm?c=Estadistica_C&cid=1254736177031&menu=ultiDatos&idp=1254734710990 codigos municipio

In [ ]:
df.columns

Index(['MES', 'DIA_SEMANA', 'HORA', 'COD_PROVINCIA', 'COD_MUNICIPIO', 'ISLA',
       'ZONA', 'CARRETERA', 'KM', 'SENTIDO_1F', 'TIPO_VIA', 'TIPO_ACCIDENTE',
       'TOTAL_HG24H', 'TOTAL_HL24H', 'TOTAL_MU30DF', 'TOTAL_HG30DF',
       'TOTAL_HL30DF', 'TOTAL_VICTIMAS_30DF', 'TOTAL_VEHICULOS',
       'TOT_PEAT_MU30DF', 'TOT_BICI_MU30DF', 'TOT_CICLO_MU30DF',
       'TOT_MOTO_MU30DF', 'TOT_TUR_MU30DF', 'TOT_FURG_MU30DF',
       'TOT_CAM_MAS3500_MU30DF', 'TOT_VMP_MU30DF', 'TOT_OTRO_MU30DF', 'NUDO',
       'PRIORI_NORMA', 'PRIORI_AGENTE', 'PRIORI_SEMAFORO', 'PRIORI_VERT_STOP',
       'PRIORI_VERT_CEDA', 'PRIORI_HORIZ_STOP', 'PRIORI_HORIZ_CEDA',
       'PRIORI_MARCAS', 'PRIORI_PEA_NO_ELEV', 'PRIORI_PEA_ELEV',
       'PRIORI_MARCA_CICLOS', 'PRIORI_CIRCUNSTANCIAL', 'PRIORI_OTRA',
       'CONDICION_NIVEL_CIRCULA', 'CONDICION_FIRME', 'CONDICION_ILUMINACION',
       'CONDICION_METEO', 'VISIB_RESTRINGIDA_POR', 'ACERA', 'TRAZADO_PLANTA',
       'TYP_ACC'],
      dtype='object')

# Ideas: 


<font color="red">Para poner mensajes en rojo </font>